# 01. Importing libraries

In [49]:
import pandas as pd
import numpy as np
import os

# 0.2 Importing Dataset

In [50]:
# Telling python to remember a main folder path
path=r'/Users/urvi/Desktop/Instacart Basket Analysis'
path

'/Users/urvi/Desktop/Instacart Basket Analysis'

In [51]:
#Importing orders_wrangled.csv data set
df_ords=pd.read_csv(os.path.join(path, '02 Data','Prepared Data','orders_wrangled.csv'), index_col=False)

In [52]:
# Importing products.csv data set
df_prods=pd.read_csv(os.path.join(path,'02 Data','Original Data', 'products.csv'), index_col=False)

In [84]:
#Importing orders_wrangled.csv data set
df_ords_clean=pd.read_csv(os.path.join(path, '02 Data','Prepared Data','orders_wrangled.csv'), index_col=False)

# 0.3 Data consistency checks

In [53]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

### Mixed Data Type

In [54]:
# Create a dataframe
df_test = pd.DataFrame()

In [55]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]
df_test.head()

mix
0     a
1     b
2     1
3  True

In [56]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird= (df_test[[col]].map(type)!=df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_test[weird])>0:
        print (col)

mix


In [57]:
# Fix mixed-type column
df_test['mix'] = df_test['mix'].astype('str')

### Missing values

#### Task 1 

In [58]:
# Function to obtain the total number of missing observations
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [59]:
# Create a subset containing only missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [60]:
df_prods.shape

(49693, 5)

In [61]:
# Create a subset containing only non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [62]:
df_prods_clean.shape

(49677, 5)

### Duplicate values

In [63]:
# Looking for full duplicates within the df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [64]:
# Removing duplicate values
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49672, 5)

# 0.4 Task Exercises

### 2. Check df_ords dataframe for inconsistencies

In [65]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

#### Missing Values: The days_since_prior_order count is less than others, indicating missing data. This impacts analysis, requiring imputation or exclusion.

#### High order_number: The maximum order_number seems unusually large, suggesting a possible data entry error or an outlier that could skew analysis.

#### Day Range: The orders_day_of_week range is 0-6, aligning with a standard programming convention for representing days, where Sunday might be 0 and Saturday 6.

#### Hour Range: The order_hour_of_day range is 0-23, reflecting a 24-hour clock format, typical for time data.

### 3. Check df_ords for mixed type data

In [66]:
for col in df_ords.columns.tolist():
    weird= (df_ords[[col]].map(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_ords[weird])>0:
        print (col)
else :
    print ("None of the columns have mix data")

None of the columns have mix data


#### No mixed-type data was found.

### 5. Run a check for missing values in your df_ords dataframe.

In [67]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### The data shows missing values exclusively in the days_since_prior_order column, with exactly 206,209 instances lacking data. This suggests that many orders are first-time purchases or the dataset did not record the interval between orders for these instances. The absence of missing values in other columns indicates that the dataset is otherwise complete, highlighting a specific issue with tracking or recording the time between orders.

### 6. Addressing appropriate method for missing value

#### Removing these missing values could eliminate valuable data on first-time orders. To address this without losing information:
#### 1. Verify that the number of missing values matches the count of order_number=1, indicating first orders.
#### 2. Add a new column to flag first orders as either True or False.
#### Every user will have one NaN value for his/her first order.  Considering all these factors it has been decided to leave the missing data untouched.

In [68]:
# Create a subset containing only missing values
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [69]:
# Check the frequency of order_number=1 in df_nan
df_nan['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

In [70]:
# Create a function to determine if the order in question is a first order or not
def is_first_order(row):
    if row['order_number'] == 1:
        return True
    else:
        return False

In [71]:
# Apply the is_first_order function to every row of the df_ords dataframe
df_ords.apply(lambda row: is_first_order(row), axis=1)

0           True
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Length: 3421083, dtype: bool

In [72]:
# Create a new column in df_ords to flag the reason for missing values
df_ords['first_order']=df_ords.apply(lambda row: is_first_order(row), axis=1)

In [73]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False

In [74]:
# Final check for the frequency of first_order=True
df_ords['first_order'].value_counts()

first_order
False    3214874
True      206209
Name: count, dtype: int64

### 7. Running duplicate check for df_ords

In [85]:
df_ords_dups = df_ords_clean[df_ords_clean.duplicated(keep=False)]
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [86]:
df_ords_dups.shape

(0, 7)

#### An empty df_ords_dups dataframe proves there are no duplicates. No action needed to remove duplicates.

# 0.5 Exporting dataframes

In [87]:
# Exporting df_prods_clean_no_dups dataframe as "products_checked.csv"
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [90]:
# Exporting df_ords dataframe as "orders_checked.csv"
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))